nix/hetzner.nix

```diff
@@ -90,12 +90,177 @@ with lib;
+        mdadm --create --run --verbose /dev/md0 --level=1 --raid-devices=2 /dev/sda2 /dev/sdb2
+        mdadm --create --run --verbose /dev/md1 --level=1 --raid-devices=2 /dev/sdc2 /dev/sdd2
+
+        # Wipe filesystem signatures that might be on the RAID from some
+        # possibly existing older use of the disks.
+        # It's not clear to me *why* it is needed, but I have certainly
+        # observed that it is needed because ext4 labels magically survive
+        # mdadm RAID re-creations.
+        # See
+        #   https://serverfault.com/questions/911370/why-does-mdadm-zero-superblock-preserve-file-system-information
+        wipefs -a /dev/md0
+        wipefs -a /dev/md1
+
+        # Disable RAID recovery. We don't want this to slow down machine provisioning
+        # in the Hetzner rescue mode. It can run in normal operation after reboot.
+        echo 0 > /proc/sys/dev/raid/speed_limit_max
+
+        # `lazy_journal_init=1` to not have to zero the device;
+        # we use ATA TRIM with RZAT support to guarantee the device
+        # is already zeroed; see comment further up about the safety of that.
+        mke2fs -F -L rootjournal -O journal_dev -E lazy_journal_init=1 /dev/md0
+        mkfs.ext4 -F -L root -J device=/dev/md0 /dev/md1
+      '';
+      description = ''
+        Script to run after booting into the Hetzner rescue mode
+        to manually create partitions.
+
+        Note as of writing, Hetzner uses BIOS, not UEFI, so if you want
+        to use GPT partition tables (which you need in case you want to
+        make partitions larger than 2 TiB) you will likely have to make
+        a BIOS boot partition
+        (<link xlink:href="http://fedoraproject.org/wiki/Anaconda/Kickstart"/>).
+
+        Where possible, use the simpler "partitions" option instead of this option.
+
+        The "partitions" and "partitioningScript" options are mutually exclusive.
+
+        If you use this option, you must set "partitions = null",
+        you must set "filesystemInfo" to an accurate representation
+        of the partitions your script creates,
+        and you must set "mountScript" to mount the created target
+        root partition at /mnt.
+      '';
+    };
+
+    mountScript = mkOption {
+      type = types.nullOr types.lines;
+      default = null;
+      example = ''
+        # Example mountScript matching the example for partitioningScript,
+        # that creates an ext4 with external journal, across two RAID1s
+        # (one on HDDs, one on SSDs).
+
+        set -e
+        mount -o data=journal /dev/md1 /mnt
+      '';
+      description = ''
+        Script to run after booting into the Hetzner rescue mode,
+        and after formatting, to mount the root filesystem at /mnt.
+
+        This option is required when "partitioningScript" is used.
+      '';
+    };
+
+    filesystemInfo = mkOption {
+      type = types.nullOr types.attrs;
+      default = null;
+      example = {
```

Comment:
Maybe use `literalExample` here, otherwise this ends up like this: 

```nix
{ boot = { loader = { grub = { devices = [ "/dev/sda" "/dev/sdb" "/dev/sdc" "/dev/sdd" ] ; } ; } ; } ; fileSystems = { / = { fsType = "ext4"; label = "root"; options = [ "journal_path=/dev/disk/by-label/rootjournal" "data=journal" "errors=remount-ro" ] ; } ; } ; swapDevices = [ ] ; }
```

Reply:
Done